test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
                             MAX_ABSDIFF)
 
 }
+
+enum TestOutputArrayType {
+    NULL_ARRAY,
+    FULL_ARRAY,
+    SUB_ARRAY,
+    REORDERED_ARRAY
+};
+
+class TestOutputArrayInfo {
+    af_array out_arr_c; // This af_array determines what will be given back to
+                        // the af_* function, and thus what will be released
+                        // after it is used
+    af::index out_subarr_idxs[4];
+    af::array subarr;
+    TestOutputArrayType out_arr_type;
+
+public:
+    // There might be a better way of encapsulating these, but for now, these
+    // are directly exposed to simplify dealing with af::array reference counts
+    af::array out_arr;
+    af::array out_arr_cpy;
+
+    TestOutputArrayInfo(TestOutputArrayType arr_type)
+        :out_arr_type(arr_type) {}
+
+    ~TestOutputArrayInfo() {
+        if (out_arr_c != 0) af_release_array(out_arr_c);
+    }
+
+    TestOutputArrayType getOutputArrayType() { return out_arr_type; }
+
+    void initArrays() {
+        out_arr_c = 0;
+    }
+
+    void initArrays(af::dim4 dims, af::dtype ty) {
+        out_arr = af::randu(dims, ty);
+        out_arr_cpy = out_arr.copy();
+        af_retain_array(&out_arr_c, out_arr.get());
+    }
+
+    void initArrays(af::dim4 dims, af::dtype ty, af::index subarr_idxs[4]) {
+        out_arr = af::randu(dims, ty);
+        out_arr_cpy = out_arr.copy();
+
+        for (int i = 0; i < 4; ++i) {
+            out_subarr_idxs[i] = subarr_idxs[i];
+        }
+        subarr = out_arr(
+            out_subarr_idxs[0],
+            out_subarr_idxs[1],
+            out_subarr_idxs[2],
+            out_subarr_idxs[3]);
+
+        // The output array that will be given to the client is the subarray,
+        // not the full array
+        af_retain_array(&out_arr_c, subarr.get());
+    }
+
+    af_array getOutputArrayC() {
+        return out_arr_c;
+    }
+
+    void setOutputArray(af::array array) {
+        // When reassigning the output array, release the old retained af_array
+        // then retain the new af_array of the new C++ array
+        af_release_array(out_arr_c);
+        out_arr = array;
+        af_retain_array(&out_arr_c, out_arr.get());
+    }
+
+    // This is used in the case of NULL_ARRAY, when taking the output array
+    // generated by the af_* function. Doing this gives TestOutputArrayInfo the
+    // responsibility to release that output array
+    void setOutputArrayC(af_array array) {
+        out_arr_c = array;
+    }
+
+    af::index getSubArrayIdx(dim_t dim) {
+        return out_subarr_idxs[dim];
+    }
+};
+
+// Generates a null af_array. testWriteToOutputArray expects that the array that
+// it receives after the af_* function is a valid, allocated af_array
+void genNullArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    metadata.initArrays();
```

Comment:
`genNullArray` doesn't need to call anything, it will be a no op function. We can keep it if we want to use keep the calling pattern inside switch case in the function `genTestOutputArray`. 

I think it should be removed since it doesn't do anything.

Reply:
Yes this is essentially a no-op. I debated whether to leave this here or not but I decided to leave it in order to keep the calling pattern in the switch statement in `genTestOutputArray`. But yeah we don't have to keep it (calling that empty function is a small performance hit too).